### DISCUSSION TITLE
Promoting myservice from staging -> production (using app of apps or applicationset)

### DISCUSSION THREAD

**Discussion by ghostsquad** [argoproj/argo-cd#5290] [Ideas]
Problem Statement:

Given an App of Apps pattern (or new ApplicationSet)

Parent App contains
- myservice-staging
- myservice-production

I want to "sync" an `myservice-staging` and only after everything is "healthy", proceed to another wave and sync `myservice-production`

Things to consider:

- How do I verify my deployment went healthy (new replicas, no crashlooping, etc)?
- If I'm using ArgoRollouts, I must ensure that the Rollout completes all steps and promotion was successful
- What about other CRDs? If there was a way to _tell_ ArgoCD about how to determine if the CRD is healthy, can I do that without contributing to the core project/waiting for a new release?
- How do I run other post-deployment verification? (maybe this is a vanilla Kubernetes Job, or maybe it's an ArgoWorkflow)

---

If this is all possible, I think there should be explicit documentation somewhere to describe how to enact this pattern.

If there are missing pieces, I'd like to know if a separate resource such as https://github.com/maruina/argocd-progressive-rollout-controller is needed, or if contributions to the core of ArgoCD would be better?

**Comment by maruina** [comment]
I think that
> What about other CRDs? If there was a way to tell ArgoCD about how to determine if the CRD is healthy, can I do that without contributing to the core project/waiting for a new release?

You can write your custom health check without waiting for a new release https://argoproj.github.io/argo-cd/operator-manual/health/#way-1-define-a-custom-health-check-in-argocd-cm-configmap

> How do I run other post-deployment verification? (maybe this is a vanilla Kubernetes Job, or maybe it's an ArgoWorkflow)

Something like flagger (https://docs.flagger.app/) would be useful. Even if you don't want to use the built-in canary analisys, you can use the webhook to perform your validation (https://docs.flagger.app/usage/webhooks)
